cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -0,0 +1,449 @@
+ * </p>
+ * <p>
+ * Pass 2: This pass only happen if a method with {@link AuthEnforce} annotation is found in the class during the
+ * first pass. In this pass we rewrite the method which has {@link AuthEnforce} annotation. In the rewrite we
+ * generate a call to
+ * {@link AuthEnforceAnnotationEnforcer#enforce(AuthorizationEnforcer, EntityId, AuthenticationContext, Set)} with
+ * all the annotation details collected in the first pass.
+ * </p>
+ * <p>
+ * Below is a sample of how a class containing a method annotation with {@link AuthEnforce} looks like before and
+ * after class rewrite.
+ * </p>
+ * <p>
+ * Before:
+ * <pre>
+ *     public class ValidAuthEnforceAnnotations {
+ *      private final AuthorizationEnforcer authorizationEnforcer;
+ *      private final AuthenticationContext authenticationContext;
+ *
+ *      public ValidAuthEnforceAnnotations() {
+ *        authenticationContext = new AuthenticationTestContext();
+ *        authorizationEnforcer = new ExceptionAuthorizationEnforcer();
+ *      }
+ *
+ *      &#064;AuthEnforce(entities = "namespaceId", enforceOn = NamespaceId.class, actions = Action.ADMIN)
+ *      public void testSingleAction(@Name("namespaceId") NamespaceId namespaceId) throws Exception {
+ *        System.out.println("Hello");
+ *      }
+ *     }
+ *   </pre>
+ * </p>
+ * <p>
+ * After:
+ * <pre>
+ *     public class ValidAuthEnforceAnnotations {
+ *      private final AuthorizationEnforcer authorizationEnforcer;
+ *      private final AuthenticationContext authenticationContext;
+ *
+ *      public ValidAuthEnforceAnnotations() {
+ *        authenticationContext = new AuthenticationTestContext();
+ *        authorizationEnforcer = new ExceptionAuthorizationEnforcer();
+ *      }
+ *
+ *      public void testSingleAction(@Name("namespaceId") NamespaceId namespaceId) throws Exception {
+ *        AuthEnforceAnnotationEnforcer.enforce(this.authorizationEnforcer, namespaceId, authenticationContext,
+ *                                              Set<Action.Admin>);
+ *        System.out.println("Hello");
+ *      }
+ *     }
+ *   </pre>
+ * </p>
+ */
+public class AuthEnforceRewriter implements ClassRewriter {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AuthEnforceRewriter.class);
+
+  private static final Type AUTHORIZATION_ENFORCER_TYPE = Type.getType(AuthorizationEnforcer.class);
+  private static final Type AUTHENTICATION_CONTEXT_TYPE = Type.getType(AuthenticationContext.class);
+  private static final Type ACTION_TYPE = Type.getType(Action.class);
+  private static final Type AUTH_ENFORCE_ANNOTATION_ENFORCER_TYPE = Type.getType(AuthEnforceAnnotationEnforcer.class);
+  private Map<Method, AnnotationDetails> methodAnnotations;
+
+  @Override
+  public byte[] rewriteClass(String className, InputStream input) throws IOException {
+    byte[] classBytes = ByteStreams.toByteArray(input);
+    methodAnnotations = new HashMap<>(); // re-initialize for this new class.
```

Comment:
Better not use field. Have a method in `AuthEnforceAnnotationVisitor` class to return this map instead.